### ISSUE TITLE
CS8602 should look deeper ("Dereference possibly null reference")

### ISSUE THREAD

**Issue by JVimes** [dotnet/roslyn#74797]
**Brief description:**

CS8602 seems to have a lot of false positives. Suggest making it look deeper to see if value could actually be null.

![2024-08-18_13h44_27](https://github.com/user-attachments/assets/9dd3bea2-729b-458e-b57e-f8054f547763)

**Languages applicable:**

C#  
Not sure on VB.

**Code example that the analyzer should [not] report:**

```c#
var text = "foo";
var notNull = text != null;
if (notNull)
{
    var length = text.Length;
}
```

**Additional information:**

I vaguely recall that the analzyers used to look deeper to see if a reference could actually be null. Didn't that used to be the case?

**Documentation requirements:**

N/A

**Comment by huoyaoyuan** [comment]
This is by design. All the nullable features only take account for expressions directly inside the `if` statement. Boolean variables are never propagated. See also #27011 #66807 .

Also note that nullable analysis isn't "analyzer". It is a part of core compiler/language and uses CSxxxx id.

**Comment by JVimes** [comment]
@Hyleon-Ou Thank you. Are requests for design changes accepted, or not really?

**Comment by huoyaoyuan** [comment]
It's unlikely to be changed, because more deep analysis may have unconstrained overhead.
